File: app/lib/models/post.dart
import 'user.dart';

class Post {
  final String id;
  final String createdAt;
  final String updatedAt;
  final String name;
  final String? description;
  final bool verified;
  final bool thingReturnedToOwner;
  final bool hasPhoto;
  final User author;

  const Post({
    required this.id,
    required this.createdAt,
    required this.updatedAt,
    required this.name,
    this.description,
    required this.verified,
    required this.thingReturnedToOwner,
    required this.hasPhoto,
    required this.author,
  });

  factory Post.fromJson(Map<String, dynamic> json) => Post(
    id: json['id'] as String,
    createdAt: json['createdAt'] as String,
    updatedAt: json['updatedAt'] as String,
    name: json['name'] as String,
    description: json['description'] as String?,
    verified: json['verified'] as bool,
    thingReturnedToOwner: json['thingReturnedToOwner'] as bool,
    hasPhoto: json['hasPhoto'] as bool,
    author: User.fromJson(json['author'] as Map<String, dynamic>),
  );
}

File: backend/internal/model/post.go
package model

import (
	"github.com/google/uuid"
	"gorm.io/gorm"
	"time"
)

// Post contains a model of a table with info about lost things in the format of
// posts
type Post struct {
	// UUID4; it will be used in URIs to see statuses of the posts
	ID          uuid.UUID `gorm:"type:uuid;default:gen_random_uuid();primaryKey"`
	CreatedAt   time.Time
	UpdatedAt   time.Time
	DeletedAt   gorm.DeletedAt
	Name        string `gorm:"type:varchar(50);check:length(trim(name)) >= 2"`
	Description string `gorm:"type:varchar(1000)"`
	// was the thing found, i.e. returned to owner? (true/false)
	ThingReturnedToOwner bool `gorm:"type:boolean;default:false"`
	// the logic is the same as for user's avatar
	HasPhoto bool      `gorm:"type:boolean;default:false"`
	AuthorID uuid.UUID `gorm:"type:uuid"`
	Author   User      `gorm:"foreignKey:AuthorID;references:ID"`
	// one-to-one (post-to-post_moderation)
	Moderation PostModeration `gorm:"foreignKey:PostID;references:ID;constraint:OnDelete:cascade,OnUpdate:restrict"`
}
